File: binary_tree/binary_tree.py
from turtle import right


class node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class binary_tree:

    def __init__(self, root):
        self.root = root

    def add_node(self, data):
        if self.root.data > data:
            if self.root.left == None:
                new_node = node(data)
                self.root.left = binary_tree(new_node)
            else:
                self.root.left.add_node(data)
        else:
            if self.root.right == None:
                new_node = node(data)
                self.root.right = binary_tree(new_node)
            else:
                self.root.right.add_node(data)

    def traverse_in_order(self, array):
        if not self.root:
            return
        if self.root.left:
            self.root.left.traverse_in_order(array)
        array.append(self.root.data)
        if self.root.right:
            self.root.right.traverse_in_order(array)
        # return array

    def preorder_traversal(self, array):
        array.append(self.root.data)
        if not self.root:
            return
        if self.root.left:
            self.root.left.preorder_traversal(array)
        if self.root.right:
            self.root.right.preorder_traversal(array)

    def postorder_traversal(self, array):
        if not self.root:
            return
        if self.root.left:
            self.root.left.postorder_traversal(array)
        if self.root.right:
            self.root.right.postorder_traversal(array)
        array.append(self.root.data)


root = node(58)

bt = binary_tree(root)

bt.add_node(2)

bt.add_node(3)
bt.add_node(44)
bt.add_node(3)
bt.add_node(25)
bt.add_node(60)
bt.add_node(7)
bt.add_node(77)
bt.add_node(9)
bt.add_node(10)


in_order = []
pre_order = []
post_order = []
bt.traverse_in_order(in_order)
bt.preorder_traversal(pre_order)
bt.postorder_traversal(post_order)
print(f'in-order:   {in_order}')
print(f'pre-order   {pre_order}')
print(f'post-order: {post_order}')

File: binary_tree/binary_tree.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct node
{
    int data;
    struct node *left;
    struct node *right;
} node;

node *head;

void insert_node(int val, node *HEAD)
{
    if (val <= (HEAD)->data)
    {
        if ((HEAD)->left == NULL)
        {
            node *new_node = (node *)malloc(sizeof(node));
            new_node->data = val;
            new_node->left = NULL;
            new_node->right = NULL;
            (HEAD)->left = new_node;
        }
        else
        {
            node *new_n = (node *)malloc(sizeof(node));
            new_n = (HEAD)->left;
            new_n->left = NULL;
            new_n->right = NULL;
            insert_node(val, new_n);
        }
    }
    else
    {
        if ((HEAD)->right == NULL)
        {
            node *new_node = (node *)malloc(sizeof(node));
            new_node->data = val;
            new_node->left = NULL;
            new_node->right = NULL;
            (HEAD)->right = new_node;
        }
        else
        {
            node *new_node = (node *)malloc(sizeof(node));
            new_node = (HEAD)->right;
            new_node->left = NULL;
            new_node->right = NULL;
            insert_node(val, new_node);
        }
    }
}

bool search(node *HEAD, int val)
{
    if (HEAD == NULL)
    {
        printf("%d not found\n", val);
        return false;
    }
    if (HEAD->data == val)
    {
        printf("%d found\n", val);
        return true;
    }
    if (HEAD->data < val)
    {
        return search(HEAD->right, val);
    }
    if (HEAD->data > val)
    {
        return search(HEAD->left, val);
    }
    printf("%d not found\n", val);
    return false;
}

void in_order_traversal(node *HEAD)
{
    if (HEAD != NULL)
    {
        in_order_traversal(HEAD->left);
        printf("%d\t", HEAD->data);
        in_order_traversal(HEAD->right);
    }
    return;
}

void pre_order_traversal(node *HEAD)
{
    if (HEAD != NULL)
    {
        printf("%d\t", HEAD->data);
        in_order_traversal(HEAD->left);
        in_order_traversal(HEAD->right);
    }
    return;
}

void post_order_traversal(node *HEAD)
{
    if (HEAD != NULL)
    {
        in_order_traversal(HEAD->left);
        in_order_traversal(HEAD->right);
        printf("%d\t", HEAD->data);
    }
    return;
}

int main()
{
    head = (node *)malloc(sizeof(node));
    head->data = 1;
    head->left = NULL;
    head->right = NULL;

    insert_node(2, head);
    insert_node(0, head);
    insert_node(-1, head);

    printf("%d\n", head->data);
    node *temp = (node *)malloc(sizeof(node));
    temp = head->right;
    printf("%d\n", temp->data);
    search(head, 5);
    in_order_traversal(head);
    printf("\n");
    pre_order_traversal(head);
    printf("\n");
    post_order_traversal(head);

    return 0;
}